File: sdk/flutter/lib/src/annotations/probe_types.dart
/// Core types for the UI Test Probe annotation system.

/// Classification of UI elements by their semantic role.
/// Matches the canonical 9-value spec in spec/probe-types.ts.
enum ProbeType {
  /// Table, DataGrid, List, Tree.
  dataContainer,

  /// ComboBox, Select, Dropdown.
  selector,

  /// Button, Link, MenuItem.
  action,

  /// Label, Badge, Counter, Chip.
  display,

  /// Image, Video, Audio, Canvas.
  media,

  /// Input, TextArea, Checkbox, Radio.
  form,

  /// Page-level container — anchor for waitForPageReady().
  page,

  /// Dialog, Drawer, Popover, Tooltip.
  modal,

  /// Tabs, Breadcrumb, Sidebar, Paginator.
  navigation,
}

/// The effect type of a linkage between probe elements.
/// Matches the canonical 7-value spec in spec/probe-types.ts.
enum LinkageEffect {
  /// Updates available options in a selector.
  optionsUpdate,

  /// Triggers data reload via API.
  dataReload,

  /// Shows or hides another element.
  visibilityToggle,

  /// Enables or disables another element.
  enabledToggle,

  /// Updates the value of another element.
  valueUpdate,

  /// Resets another element to default state.
  reset,

  /// Triggers navigation to a different route.
  navigate,
}

/// Describes a linkage path from one probe element to another.
class LinkagePath {
  /// The probe ID of the target element.
  final String targetId;

  /// The effect this linkage produces.
  final LinkageEffect effect;

  /// Optional condition that must be met for this linkage to activate.
  final String? condition;

  const LinkagePath({
    required this.targetId,
    required this.effect,
    this.condition,
  });

  @override
  String toString() => 'LinkagePath($targetId, $effect)';
}

/// Describes the device profile for cross-device testing.
class DeviceProfile {
  /// Device name (e.g., "iPhone 15 Pro", "Pixel 8").
  final String name;

  /// Logical screen width in dp/pt.
  final double width;

  /// Logical screen height in dp/pt.
  final double height;

  /// Device pixel ratio.
  final double pixelRatio;

  /// Platform identifier (ios, android, web).
  final String platform;

  const DeviceProfile({
    required this.name,
    required this.width,
    required this.height,
    this.pixelRatio = 1.0,
    required this.platform,
  });
}

/// Platform context information captured at scan time.
class PlatformContext {
  /// The platform (ios, android, web, macos, linux, windows).
  final String platform;

  /// OS version string.
  final String osVersion;

  /// Device profile if available.
  final DeviceProfile? device;

  /// Locale identifier (e.g., "en_US").
  final String locale;

  /// Whether dark mode is active.
  final bool isDarkMode;

  /// Text scale factor.
  final double textScaleFactor;

  const PlatformContext({
    required this.platform,
    required this.osVersion,
    this.device,
    required this.locale,
    this.isDarkMode = false,
    this.textScaleFactor = 1.0,
  });
}

/// A fully-resolved probe element with all 15 fields from the spec.
class ProbeElement {
  /// Unique probe identifier.
  final String id;

  /// Semantic type classification.
  final ProbeType type;

  /// Human-readable label.
  final String? label;

  /// Current value (for inputs/displays).
  final dynamic value;

  /// Semantic state map (e.g., {"enabled": true, "loading": false}).
  final Map<String, dynamic> state;

  /// Data source URI or identifier.
  final String? source;

  /// Linkage paths to other probe elements.
  final List<LinkagePath> linkage;

  /// Bounding box: [x, y, width, height] in logical pixels.
  final List<double>? bounds;

  /// Whether the element is effectively visible on screen.
  final bool isVisible;

  /// Whether the element is enabled for interaction.
  final bool isEnabled;

  /// The screen/route this element belongs to.
  final String? screen;

  /// Accessibility label.
  final String? a11yLabel;

  /// Accessibility role.
  final String? a11yRole;

  /// Parent probe ID in the hierarchy.
  final String? parentId;

  /// Platform context at capture time.
  final PlatformContext? platformContext;

  const ProbeElement({
    required this.id,
    required this.type,
    this.label,
    this.value,
    this.state = const {},
    this.source,
    this.linkage = const [],
    this.bounds,
    this.isVisible = true,
    this.isEnabled = true,
    this.screen,
    this.a11yLabel,
    this.a11yRole,
    this.parentId,
    this.platformContext,
  });

  @override
  String toString() => 'ProbeElement($id, $type)';
}

File: sdk/ios/Sources/UITestProbe/Models/ProbeTypes.swift
import Foundation

/// Classification of UI elements by their semantic role.
/// Matches the canonical 9-value spec in spec/probe-types.ts.
public enum ProbeType: String, Codable, Sendable {
    /// Table, DataGrid, List, Tree.
    case dataContainer = "data-container"
    /// ComboBox, Select, Dropdown.
    case selector
    /// Button, Link, MenuItem.
    case action
    /// Label, Badge, Counter, Chip.
    case display
    /// Image, Video, Audio, Canvas.
    case media
    /// Input, TextArea, Checkbox, Radio.
    case form
    /// Page-level container — anchor for waitForPageReady().
    case page
    /// Dialog, Drawer, Popover, Tooltip.
    case modal
    /// Tabs, Breadcrumb, Sidebar, Paginator.
    case navigation
}

/// The effect type of a linkage between probe elements.
/// Matches the canonical 7-value spec in spec/probe-types.ts.
public enum LinkageEffect: String, Codable, Sendable {
    /// Updates available options in a selector.
    case optionsUpdate = "options_update"
    /// Triggers data reload via API.
    case dataReload = "data_reload"
    /// Shows or hides another element.
    case visibilityToggle = "visibility_toggle"
    /// Enables or disables another element.
    case enabledToggle = "enabled_toggle"
    /// Updates the value of another element.
    case valueUpdate = "value_update"
    /// Resets another element to default state.
    case reset
    /// Triggers navigation to a different route.
    case navigate
}

/// Describes a linkage path from one probe element to another.
public struct LinkagePath: Codable, Sendable {
    /// The probe ID of the target element.
    public let targetId: String
    /// The effect this linkage produces.
    public let effect: LinkageEffect
    /// Optional condition that must be met for this linkage to activate.
    public let condition: String?

    public init(targetId: String, effect: LinkageEffect, condition: String? = nil) {
        self.targetId = targetId
        self.effect = effect
        self.condition = condition
    }
}

/// Describes the device profile for cross-device testing.
public struct DeviceProfile: Codable, Sendable {
    /// Device name (e.g., "iPhone 15 Pro", "iPad Air").
    public let name: String
    /// Logical screen width in points.
    public let width: Double
    /// Logical screen height in points.
    public let height: Double
    /// Device pixel ratio (scale factor).
    public let pixelRatio: Double
    /// Platform identifier ("ios", "macos").
    public let platform: String

    public init(name: String, width: Double, height: Double, pixelRatio: Double = 1.0, platform: String) {
        self.name = name
        self.width = width
        self.height = height
        self.pixelRatio = pixelRatio
        self.platform = platform
    }
}

/// Platform context information captured at scan time.
public struct PlatformContext: Codable, Sendable {
    /// The platform ("ios", "macos").
    public let platform: String
    /// OS version string.
    public let osVersion: String
    /// Device profile if available.
    public let device: DeviceProfile?
    /// Locale identifier (e.g., "en_US").
    public let locale: String
    /// Whether dark mode is active.
    public let isDarkMode: Bool
    /// Text scale factor (Dynamic Type).
    public let textScaleFactor: Double

    public init(
        platform: String,
        osVersion: String,
        device: DeviceProfile? = nil,
        locale: String,
        isDarkMode: Bool = false,
        textScaleFactor: Double = 1.0
    ) {
        self.platform = platform
        self.osVersion = osVersion
        self.device = device
        self.locale = locale
        self.isDarkMode = isDarkMode
        self.textScaleFactor = textScaleFactor
    }
}

/// A fully-resolved probe element with all 15 fields from the spec.
public struct ProbeElement: Codable, Sendable {
    /// Unique probe identifier.
    public let id: String
    /// Semantic type classification.
    public let type: ProbeType
    /// Human-readable label.
    public let label: String?
    /// Current value (encoded as String for Codable compliance).
    public let value: String?
    /// Semantic state map (e.g., ["enabled": "true", "loading": "false"]).
    public let state: [String: String]
    /// Data source URI or identifier.
    public let source: String?
    /// Linkage paths to other probe elements.
    public let linkage: [LinkagePath]
    /// Bounding box: [x, y, width, height] in points.
    public let bounds: [Double]?
    /// Whether the element is effectively visible on screen.
    public let isVisible: Bool
    /// Whether the element is enabled for interaction.
    public let isEnabled: Bool
    /// The screen/route this element belongs to.
    public let screen: String?
    /// Accessibility label.
    public let a11yLabel: String?
    /// Accessibility role.
    public let a11yRole: String?
    /// Parent probe ID in the hierarchy.
    public let parentId: String?
    /// Platform context at capture time.
    public let platformContext: PlatformContext?

    public init(
        id: String,
        type: ProbeType,
        label: String? = nil,
        value: String? = nil,
        state: [String: String] = [:],
        source: String? = nil,
        linkage: [LinkagePath] = [],
        bounds: [Double]? = nil,
        isVisible: Bool = true,
        isEnabled: Bool = true,
        screen: String? = nil,
        a11yLabel: String? = nil,
        a11yRole: String? = nil,
        parentId: String? = nil,
        platformContext: PlatformContext? = nil
    ) {
        self.id = id
        self.type = type
        self.label = label
        self.value = value
        self.state = state
        self.source = source
        self.linkage = linkage
        self.bounds = bounds
        self.isVisible = isVisible
        self.isEnabled = isEnabled
        self.screen = screen
        self.a11yLabel = a11yLabel
        self.a11yRole = a11yRole
        self.parentId = parentId
        self.platformContext = platformContext
    }
}
